backend/src/main/java/com/example/backend/Specifications/FlightSpecifications.java

```diff
@@ -23,20 +27,19 @@ public static Specification<Flight> containsSource(String source) {
 
     public static Specification<Flight> containsDestination(String destination) {
         return (root, query, criteriaBuilder) -> {
-            Subquery<Integer> subquery = query.subquery(Integer.class);
-            Root<FlightLeg> subRoot = subquery.from(FlightLeg.class);
-    
-            subquery.select(criteriaBuilder.greatest(subRoot.get("flightLegId")))
-                    .where(criteriaBuilder.equal(subRoot.get("flight"), root))
-                    .groupBy(subRoot.get("flight"));
-    
+
+            if (query == null) {
+                throw new IllegalArgumentException("Query parameter cannot be null");
+            }
+
+            Subquery<Integer> subquery = createNumberOfLastFlightLegSubquery(root, query, criteriaBuilder);
+
             Join<Flight, FlightLeg> flightAndFlightLeg = root.join("flightLegs");
             Join<FlightLeg, Airport> flightLegAndAirport = flightAndFlightLeg.join("arrivalAirport");
-    
+
             return criteriaBuilder.and(
-                criteriaBuilder.equal(flightAndFlightLeg.get("flightLegId"), subquery),
-                criteriaBuilder.equal(flightLegAndAirport.get("airportCity"), destination)
-            );
+                    criteriaBuilder.equal(flightAndFlightLeg.get("flightLegId"), subquery),
+                    criteriaBuilder.equal(flightLegAndAirport.get("airportCity"), destination));
         };
     }
```

Comment:
Here the containsSource and containsDestination functions, they could be one function with an extra parameter to avoid duplicated code.

Reply:
Same as the previous comment. But you make me notice that both containSource and hasDepartureAirport are the same except that the type of the final attribute. Same for destination. So I fix it and make the code cleaner.